File: LeetCode/Problems/Python/207. Course Schedule.py
class Solution:
    
    def dfs(self, graph, i):
        #found a cycle
        if self.visited[i] == -1:
            return False
        #we finished it before
        if self.visited[i] == 1:
            return True
        
        self.visited[i] = -1
        for neighbor in graph[i]:
            if not self.dfs(graph,neighbor):
                return False
        self.visited[i] = 1
        return True
    
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        
        #intuition:
        #finding the cycle in the dependency
        from collections import defaultdict
        graph = defaultdict(list)
        for p in prerequisites:
            nextCourse, prevCourse = p[0], p[1]
            graph[prevCourse].append(nextCourse)
        
        self.visited = [0] * numCourses
        
        for c in range(numCourses):
            if not self.dfs(graph, c):
                return False
        return True

 #topological sort      
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        if numCourses == 0:
            return False
        in_degree = Counter()
        for i in range(numCourses):
            in_degree[i] = 0
            
        adj_list = defaultdict(set)
        for p in prerequisites:
            if p[0] not in adj_list[p[1]]:
                adj_list[p[1]].add(p[0])
                in_degree[p[0]]+=1

        deq = deque()
        for i in in_degree:
            if in_degree[i] == 0:
                deq.append(i)
        result = []
        while deq:
            curr = deq.popleft()
            result.append(curr)
            for adj in adj_list[curr]:
                in_degree[adj]-=1
                if in_degree[adj]==0:
     
                    deq.append(adj)
                    

        
        return len(result) == numCourses

File: LeetCode/Problems/C++/207. Course Schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        
        
        unordered_map<int, int> depend;
        unordered_map<int, unordered_set<int>> adj_list;
        queue<int> myqueue;
        vector<int> result;
        //initialize
        for(int i=0;i<numCourses;i++){
            depend[i] = 0;
        }
        for(int i=0;i<prerequisites.size();i++){
            depend[prerequisites[i][0]]++;
            adj_list[prerequisites[i][1]].insert(prerequisites[i][0]);    
        }
    
        for(auto it=depend.begin();it!=depend.end();++it){
            if(it->second == 0) myqueue.push(it->first);
        }
        
        while(!myqueue.empty()){
            int curr = myqueue.front();
            myqueue.pop();
            result.push_back(curr);
            for(auto it = adj_list[curr].begin();it!=adj_list[curr].end(); ++it){
                depend[*it]--;
                if(depend[*it]==0)myqueue.push(*it);
            }
        } 
        return result.size() == numCourses;
    }
    
};